### PR TITLE
[#1654] Fix `VaOptionList`

### LINKED ISSUE #1654: All options disabled in `VaOptionList`
**Vuestic-ui version:** 1.3.5

**Steps to reproduce**
open both:
https://vuestic.dev/en/ui-elements/option-list#default
http://vuestic-ui-demo.sub.asva.by/#/demo/va-option-list/VaOptionList.demo.vue

**What is the expected behavior?**
- options in examples shouldn't be disabled

**What is the current behavior?**
- options disabled

**Other information**

### PR DESCRIPTION
**Pull request by RVitaly1978**
## Description
close #1654 

changes:
- [x] fix `disabled` state
- [x] fix `ref`
- [x] fix initial value (`v-model`) when its type is not array

chrome-capture

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

### CODE REVIEW

File: packages/ui/src/components/va-option-list/VaOptionList.vue [hunk 1]
```diff
@@ -107,13 +100,12 @@ export default defineComponent({
 
   setup (props, { emit }) {
     const { valueComputed } = useStateful(props, emit)
+
     const { getValue, getText, getTrackBy, getDisabled } = useSelectableList(props)
 
-    const input = ref<HTMLElement>()
+    const inputs = ref<(HTMLInputElement)[]>([])
```
Comment (LighthouseKeeperYN) [inline packages/ui/src/components/va-option-list/VaOptionList.vue:106]:
```suggestion
    const inputs = ref<HTMLInputElement[]>([])
```

File: packages/ui/src/components/va-option-list/VaOptionList.vue [hunk 2]
```diff
@@ -150,8 +142,16 @@ export default defineComponent({
 
     const { computedError, computedErrorMessages } = useValidation(props, emit, reset, focus)
 
+    const computedProps = computed(() => ({
+      ...pick(props, ['name', 'color', 'readonly', 'leftLabel']),
+    }))
```
Comment (LighthouseKeeperYN) [inline packages/ui/src/components/va-option-list/VaOptionList.vue:147]:
```suggestion
    const computedProps = computed(() => pick(props, ['name', 'color', 'readonly', 'leftLabel']))
```

**Review by LighthouseKeeperYN — APPROVED** [review]

File: packages/ui/src/composables/useSelectableList.ts [hunk 3]
```diff
@@ -31,18 +27,18 @@ export function useSelectableList (props: ExtractPropTypes<typeof useSelectableL
...
+
   const getOptionByValue = (value: SelectableOption) => {
     if (!props.valueBy) { return value }
-
     return props.options.find((option: SelectableOption) => value === getValue(option)) || value
   }
+
   const getText = (option: SelectableOption): string => getOptionProperty(option, props.textBy)
-  const getDisabled = (option: SelectableOption): boolean => getOptionProperty(option, props.disabledBy)
+  const getDisabled = (option: SelectableOption): boolean => !isStringOrNumber(option) && getProp(option, props.disabledBy)
```
Comment (m0ksem) [inline packages/ui/src/composables/useSelectableList.ts:37]: Just a question. I don't really understand what is going on here.

Is it shorthand for `option !== string && option !== 0`?

Looks like getProp returns.... undefined whet option is string. Can this be handled in getProp?

```ts
  if (typeof option !== 'object') { return option }
  if (!prop || !option) { return option }
  if (typeof prop === 'string') { return getValueByPath(option, prop) }
  if (typeof prop === 'function') { return prop(option) }
  return option
```

File: packages/ui/src/components/va-option-list/VaOptionList.vue [hunk 4]
```diff
@@ -150,8 +142,16 @@ export default defineComponent({
...
+
+    const addRef = (el: HTMLInputElement, index: number) => {
```
Comment (m0ksem) [inline packages/ui/src/components/va-option-list/VaOptionList.vue:149]: Take a look at `packages/ui/src/composables/useArrayRefs.ts`

File: packages/ui/src/composables/useSelectable.ts [hunk 5]
```diff
@@ -100,7 +100,7 @@ export const useSelectable = (
       if (!props.modelValue) {
         emit('update:modelValue', [props.arrayValue])
       } else if (!Array.isArray(props.modelValue)) {
-        return
+        emit('update:modelValue', props.modelValue === props.arrayValue ? [] : [props.modelValue, props.arrayValue])
```
Comment (m0ksem) [inline packages/ui/src/composables/useSelectable.ts:103]: I can't find arrayValue in our docs. What is it?

File: packages/ui/src/components/va-option-list/VaOptionList.vue [hunk 6]
```diff
@@ -150,8 +142,16 @@ export default defineComponent({
...
+      inputs.value[index] = el
+    }
+
     onMounted(() => {
-      if (!valueComputed.value && props.defaultValue) {
+      if (valueComputed.value === undefined && props.defaultValue) {
         selectedValue.value = props.defaultValue
```
Comment (m0ksem) [inline packages/ui/src/components/va-option-list/VaOptionList.vue:155]: Looks like useStateful must be used here...

**Review by m0ksem — CHANGES_REQUESTED** [review]

File: packages/ui/src/components/va-option-list/VaOptionList.vue (hunk 4 above)
Comment (RVitaly1978) [reply packages/ui/src/components/va-option-list/VaOptionList.vue:149]: Thank you. Fine! Didn't notice that we have such a hook!

File: packages/ui/src/composables/useSelectable.ts [hunk 7]
```diff
@@ -100,7 +100,7 @@ export const useSelectable = (
       if (!props.modelValue) {
         emit('update:modelValue', [props.arrayValue])
       } else if (!Array.isArray(props.modelValue)) {
-        return
+        emit('update:modelValue', [props.modelValue, props.arrayValue])
```
[the remark this answers is not in the document]
Comment (RVitaly1978) [reply packages/ui/src/composables/useSelectable.ts]: https://vuestic.dev/en/ui-elements/checkbox
https://vuestic.dev/en/ui-elements/switch
This property is used to group input fields (`VaCheckbox`, `VaSwitch`)

File: packages/ui/src/composables/useSelectable.ts (hunk 5 above)
Comment (RVitaly1978) [reply packages/ui/src/composables/useSelectable.ts:103]: https://vuestic.dev/en/ui-elements/checkbox
https://vuestic.dev/en/ui-elements/switch
This property is used to group input fields (`VaCheckbox`, `VaSwitch`)

File: packages/ui/src/composables/useSelectableList.ts (hunk 3 above)
Comment (RVitaly1978) [reply packages/ui/src/composables/useSelectableList.ts]: Updated `getProp` and renamed to `getValueByKey`

File: packages/ui/src/components/va-option-list/VaOptionList.vue (hunk 6 above)
Comment (RVitaly1978) [reply packages/ui/src/components/va-option-list/VaOptionList.vue:150]: Removed `defaultValue` prop

**Review by rustem-nasyrov — APPROVED** [review]

File: packages/ui/src/components/va-option-list/VaOptionList.vue [hunk 8]
```diff
@@ -150,9 +131,11 @@ export default defineComponent({
 
     const { computedError, computedErrorMessages } = useValidation(props, emit, reset, focus)
 
+    const computedProps = computed(() => pick(props, ['name', 'color', 'readonly', 'leftLabel']))
+
     onMounted(() => {
-      if (!valueComputed.value && props.defaultValue) {
-        selectedValue.value = props.defaultValue
+      if (props.type !== 'radio' && !Array.isArray(props.modelValue) && process.env.NODE_ENV !== 'production') {
```
Comment (m0ksem) [inline packages/ui/src/components/va-option-list/VaOptionList.vue:137]:
```suggestion
      if (process.env.NODE_ENV !== 'production' && props.type !== 'radio' && !Array.isArray(props.modelValue)') {
```

even better create a function isDev somewhere in utils

```suggestion
      if (isDev && props.type !== 'radio' && !Array.isArray(props.modelValue)) {
```

File: packages/ui/src/services/utils.ts [hunk 9]
```diff
@@ -75,10 +75,13 @@ export const getValueByPath = <T extends Record<string, unknown>>(option: T, pro
...
  * @param option - Object to look properties inside
- * @param prop - string or function used to find nested property
+ * @param prop - string, number or function used to find nested property
  */
-export const getProp = <T extends (Record<string, unknown> | string)> (option: T, prop: string | ((t: T) => any) | Function): any => {
-  if (typeof option === 'string') { return }
+export const getValueByKey = <T extends (Record<string, unknown> | string | number)> (
+  option: T,
+  prop: string | ((t: T) => any),
+): any => {
```
Comment (m0ksem) [inline packages/ui/src/services/utils.ts:83]:
```suggestion
export const getValueByKey = <R extends unknown, T extends (Record<string, R> | string | number)> (
  option: T,
  prop: string | ((t: T) => R),
): R=> {
```

Let's have return type annotated.

**Review by m0ksem — APPROVED** [review]

**Pull request merged**